### PR TITLE
Use docker buildx for the build-image.

### PR DESCRIPTION
**Pull request by vinayakankugoyal**
#### What type of PR is this?
/kind cleanup

#### What this PR does / why we need it:
Switch to using docker buildx for the build-image. We want to consistently use buildx everywhere as highlighted in #98908.

#### Which issue(s) this PR fixes:

Fixes #

#### Does this PR introduce a user-facing change?

```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs

```

### CODE REVIEW

File: build/common.sh
Comment (vinayakankugoyal) [inline build/common.sh:387]: I can't seem to figure out a way to move DOCKER_CLI_EXPERIMENTAL=enabled to line 383, I tried setting line 383 to 
```
  local -ra build_cmd=("DOCKER_CLI_EXPERIMENTAL=enabled" "${DOCKER[@]}" buildx build --load -t "${image}" "--pull=${pull}" "${build_args}" "${context_dir}")
```
and leaving line 387 unchanged but I get the following error: 
```
build/../build/common.sh: line 387: DOCKER_CLI_EXPERIMENTAL=enabled: command not found
To retry manually, run:

DOCKER_CLI_EXPERIMENTAL=enabled DOCKER_CLI_EXPERIMENTAL=enabled docker buildx build --load -t kube-build:build-7ac69468fc-5-v1.15.8-1 --pull=false --build-arg=KUBE_BUILD_IMAGE_CROSS_TAG=v1.15.8-1 /<redacted>/k8s.io/kubernetes/_output/images/kube-build:build-7ac69468fc-5-v1.15.8-1

!!! [0214 12:56:06] Call tree:
!!! [0214 12:56:06]  1: build/release.sh:35 kube::build::build_image(...)
make: *** [Makefile:452: quick-release] Error 1
```

**Comment by BenTheElder** [comment]
We need to be ensuring a buildx builder exists, the first point in that issue needs fixing before this one.

**Comment by vinayakankugoyal** [comment]
Hmm.. maybe I misunderstood the issue. Aren't the server images built inside the build-image container? So wouldn't we want to make sure qemu is initialized before this:- https://github.com/kubernetes/kubernetes/blob/f81235605aa3904ab0d68cb5ebaf5488787b9b0c/build/lib/release.sh#L376

File: build/common.sh
Comment (BenTheElder) [inline build/common.sh:188]: This is going to be a pain in bash I think, but if we renamed this method like `is_current_builder_missing_target_arch` we could cross-check `KUBE_BUILD_PLATFORMS` and avoid doing anything in the case like we are building for x86 on x86.

We might not even care to cross check if we document the method well, I'm not sure we care about the case of targeting a non-server platform on a different arch than the host while only building server for the host arch 🤔

File: build/common.sh
Comment (BenTheElder) [inline build/common.sh:152]: let's leave this for another PR, but noting now:

I think we should probably consider not depending on `--use` in the future. We should also consider trying to avoid the qemu call when it's already setup and/or re-using the kube-multiarch-builder if it already exists.

This is a reasonable start though, and I want to avoid blocking the work depending on this PR from optimizing it further. We mostly care in the case that builds are running concurrently on a host (stop clobbering binfmt_misc unnecessarily since it's global).

Comment (vinayakankugoyal) [reply build/common.sh:152]: > I think we should probably consider not depending on --use in the future.
Is this preferred: 
```
docker buildx create --name=kube-multiarch-builder
docker buildx use kube-multiarch-builder
```

**Comment by fejta** [comment]
/unassign
LGTM

File: build/common.sh
Comment (BenTheElder) [reply build/common.sh:152]: Er sorry, I mean depend on it being the current global default, not the flag vs the command. We can pass `--builder` to all of the `docker buildx build` calls _if_ we can settle on a name consistently 🙃 

That's a larger refactor we should consider in the future, we already depend on this consistently today, and when we do it we want to consider the other points re: not recreating it unnecessarily (and thus perhaps breaking a call assuming it is available).

File: build/common.sh
Comment (BenTheElder) [reply build/common.sh:412]: we can `export DOCKER_CLI_EXPERIMENTAL=enabled` this in common.sh probably, I don't think we have a huge reason not to. Some of our makefiles just do this

File: build/common.sh
Comment (BenTheElder) [reply build/common.sh:152]: https://github.com/kubernetes/kubernetes/pull/99080/files#r579393246 is the first more interesting part of that

File: build/lib/release.sh
Comment (BenTheElder) [inline build/lib/release.sh:245]: this one doesn't match the builder names below.

we should probably dry this out into a variable.

Comment (vinayakankugoyal) [reply build/lib/release.sh]: Done.

File: build/lib/release.sh
Comment (BenTheElder) [inline build/lib/release.sh:243]: we're *not* OK with `docker run --privileged foo:tag` in this repo

also in retrospect: the build image is not one I'd expect to be multi-arch anyhow?
for the other images we'll need this, but currently the pattern is indeed that script instead https://cs.k8s.io/?q=third_party%2Fmultiarch%2F&i=nope&files=&excludeFiles=&repos=kubernetes/kubernetes

Comment (BenTheElder) [reply build/lib/release.sh:243]: they also don't have to be installed on the host https://github.com/kubernetes/kubernetes/blob/186f934e4ccb23e7dca03cc836f11f89e7d99e08/test/images/image-util.sh#L169

Comment (vinayakankugoyal) [reply build/lib/release.sh:243]: 1. Yup we don't need multi-arch for build-image. Updated the PR description.
2. When using 
${sudo} "${KUBE_ROOT}/third_party/multiarch/qemu-user-static/register/register.sh" --reset --persistent yes
and then creating a builder using 
only creates a builder for: linux/amd64, linux/386
3. Why is downloading a tar safe (https://github.com/multiarch/qemu-user-static/releases/download/"${QEMUVERSION}"/x86_64_qemu-"${QEMUARCHS[$arch]}"-static.tar.gz)?

Comment (vinayakankugoyal) [reply build/lib/release.sh:243]: Oh I didn't know that they don't have to be installed on the host. I reverted it to use `third_party/multiarch/qemu-user-static/register/register.sh" --reset --persistent yes`

Comment (vinayakankugoyal) [reply build/lib/release.sh]: Discussed offline, Ben found a way to unblock kube-apiserver without the need for qemu: https://github.com/kubernetes/kubernetes/pull/96134#discussion_r580704333

**Comment by spiffxp** [comment]
/hold
I don't think we want to merge this until buildkit has been updated, see https://github.com/kubernetes/test-infra/issues/20884#issuecomment-783840975

**Comment by BenTheElder** [comment]
> I don't think we want to merge this until buildkit has been updated, see kubernetes/test-infra#20884 (comment)

@spiffxp we're *not* pushing with buildkit. we're building images one by one for each platform (only the host platform here) and then `--load` ing them into the current docker.

This image is also never pushed (it's the `kube-build` image in the local make build process), but even for the core images we *do* push, the build doesn't push them directly, they bounce through the local dockerd first. The problem in that issue is building a multi-arch image with `--push` and having buildkit do the pushing of the whole manifest image.

**Comment by BenTheElder** [comment]
see https://github.com/kubernetes/kubernetes/pull/99080/files#diff-64833c91e011032ba2c7a5d17a52dd97ff3d10ecc644ffbfad15259718c75813R383 for `--load` (not `--push`)

**Comment by spiffxp** [comment]
>This image is also never pushed

/hold cancel
SGTM then

**Comment by pacoxu** [comment]
1.21 or 1.22 ? 

need rebase for conflicts.

**Comment by vinayakankugoyal** [comment]
> 1.21 or 1.22 ?

Since code freeze for 1.21 is already in effect, 1.22.

**Comment by vinayakankugoyal** [comment]
> need rebase for conflicts.

Thanks Done!

**Comment by vinayakankugoyal** [comment]
@BenTheElder - do we wan't to make this change? I believe we are using buildx for docker builds for other images already.

**Review by BenTheElder — APPROVED** [review]
LGTM

buildx is already a hard requirement for building a Kubernetes release since at least v1.21.0

**Comment by BenTheElder** [comment]
/test pull-kubernetes-node-e2e-containerd

[new job that is stuck pending because it was added while this PR was inactive]

**Comment by vinayakankugoyal** [comment]
pull-kubernetes-node-e2e-containerd passed, @BenTheElder can you  /lgtm when you get a chance? Thanks!

**Comment by BenTheElder** [comment]
bah, in some repos github approve is lgtm + approve with prow
/lgtm
/approve

**Pull request merged**
